### PR TITLE
Integral sensitivity in FluxPointsEstimator

### PR DESCRIPTION
**Pull request by AtreyeeS**
The PR address #4329

The issue is how to keep backwards compatibility while renaming `energy` to `e_min` in the `Table` column.
One option might be to have both columns in the table, but that is rather ugly.

### CODE REVIEW

**Comment by adonath** [comment]
Thanks a lot @AtreyeeS! Indeed to get a single value for the integral sensitivity we can just use a single energy bin. In general (and also requested in #4329) I think it would be good to have this for an array of minimum energies.  

Concerning this PR:
- I think adding the `e_min` and `e_max` to the table is a very good idea. For backwards compatibility we could keep `energy` as well. But we should probably emit a deprecation warning in the `.run()` method.
- I think the example with a single energy bins is useful, but probably not as useful as having the same information fo an array of energies.

So to fully address the issue we would need a few more modifications (but this can be a separate PR):

- Adapt the `SensitivityEstimator` to compute the sensitive for the energy bins defined e.g. by `e_min = [1, 2, 3] * u.TeV` and `e_max = [np.inf, np.inf, np.inf] * u.TeV`. I think the definition of an `e_ref` would not apply here.
- Currently the `SensitivityEstimator` just takes the energy binning from the dataset. The `SensitivityEstimator`could be modified to take `energy_min` and `energy_max` and `energy` arguments separately. For the integral sensitivity one would then explicitly use `SensitivityEstimator(energy_min=[1, 2, 3] * u.TeV, energy_max=np.inf * u.TeV)`
- In this case one would internally have a loop of energy bins and resample the dataset to the given energy range.
- Alternatively we could just keep the logic of  taking the energy binning from the dataset.
- In this case it could be very useful to have a method like `CountsStatistic.cumsum()`, which cumulates the counts along the given axis. Maybe this is worth an independent PR.

**Comment by AtreyeeS** [comment]
Is it very useful to have the integral sensitivity as a function of energy? Wouldn't it be more practical then to compute a differential sensitivity?

**Comment by adonath** [comment]
@AtreyeeS I think this is useful in the context of source detection and finding the energy threshold which corresponds to the best sensitivity. But maybe @STSpencer can comment again.

**Comment by STSpencer** [comment]
Thanks for working on this @AtreyeeS !

So the reason we want this feature is to determine whether or not particular molecular clouds with various emission scenarios will be visible with CTA, in a similar manner to Mitchell et al. 2021 (https://arxiv.org/abs/2103.01787). For this you need to compute the predicted integrated flux from the source to the integrated sensitivity above a given energy threshold. In principle this is just one number per threshold, but I think @adonath is right (I should say I'm not an expert in this) in that being able to compute this number for arrays of minimum energy thresholds is the way to go, as this seems to be important for transient studies (see (Fioretti et al. 2016) https://arxiv.org/abs/1608.04992).

**Comment by AtreyeeS** [comment]
Thanks! Indeed, I see now. 
Maybe we merge this PR now for 1.1, so that users can at least manually loop over thresholds and get the integral sensitivity curve. We keep #4329 open for 1.2, to be implemented with @adonath suggestions.

**Review by QRemy — DISMISSED** [review]
Thanks @AtreyeeS, looks good,  no further comments from my side. I agree the remaining comments can be implemented in a separate PR.

File: gammapy/estimators/points/sensitivity.py [hunk 1]
```diff
@@ -135,6 +141,11 @@ def run(self, dataset):
         criterion = self._get_criterion(
             excess.data.squeeze(), dataset.background.data.squeeze()
         )
+        logging.warning(
```
Comment (registerrier) [inline gammapy/estimators/points/sensitivity.py:144]: Does this require a deprecation warning? It is a change in the output format, not in the API.

File: gammapy/estimators/points/tests/test_sensitivity.py [hunk 2]
```diff
@@ -41,28 +41,63 @@ def test_cta_sensitivity_estimator(spectrum_dataset):
     sens = SensitivityEstimator(gamma_min=25, bkg_syst_fraction=0.075)
     table = sens.run(dataset_on_off)
 
+    warning_message = (
```
Comment (registerrier) [inline gammapy/estimators/points/tests/test_sensitivity.py:44]: I am not sure you should test for this.

File: examples/tutorials/analysis-1d/cta_sensitivity.py [hunk 3]
```diff
@@ -188,20 +189,52 @@
...
+######################################################################
+# Obtaining an integral flux sensitivity
+# --------------------------------------
+#
+# It is often useful to obtain the integral sensitivity above a certain
+# threshold. In this case, it is simplest to use a dataset with one energy bin
+# while setting the high energy edge to a very large value
+#
+
+energy_axis1 = MapAxis.from_energy_bounds("0.03 TeV", "300 TeV", nbin=1)
```
Comment (registerrier) [inline examples/tutorials/analysis-1d/cta_sensitivity.py:215]: here you could squash the `energy_axis` defined above

Comment (registerrier) [reply examples/tutorials/analysis-1d/cta_sensitivity.py:215]: Or even simpler `geom.squash("energy")`

File: examples/tutorials/analysis-1d/cta_sensitivity.py [hunk 4]
```diff
@@ -188,20 +189,52 @@
...
+geom = RegionGeom.create("icrs;circle(0, 0.5, 0.1)", axes=[energy_axis1])
+empty_dataset1 = SpectrumDataset.create(geom=geom, energy_axis_true=energy_axis_true)
```
Comment (registerrier) [inline examples/tutorials/analysis-1d/cta_sensitivity.py:217]: or `dataset_on_off.to_image()` or `dataset.resample_energy_axis`

**Review by registerrier — COMMENTED** [review]
Thanks @AtreyeeS . This looks good.

I have left some inline comments to simplify the reduction of the dataset in the notebook.
Also I don't think we need deprecation here. You change the output table format, this is different.

A question might be whether or not to use the `FluxPoints` object instead of a table directly.

File: gammapy/estimators/points/sensitivity.py (hunk 1 above)
Comment (AtreyeeS) [reply gammapy/estimators/points/sensitivity.py:144]: Thanks @registerrier ! Interpreting backwards compatibility loosely, I tend to see it as "something which breaks my script which used to work on a prev version". In this case, users might like to be fore-warned that `table["energy"]` would yield errors in future versions?

**Comment by bkhelifi** [comment]
I would have expected in the tutorial that the integral sensitivity is similar to the differential one, ie the minimal flux above an energy threshold as a function of threshold. The standard output should look like the differential sensitivity, ie a plot with X=energy and Y=flux [/cm2/s] .
With this implementation, one needs to re-run n times the `SensitivityEstimator`. It seems to me not very efficient, but maybe we have no choice. At least one should add the parallel processing for the loop on energy thresholds

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS . This looks good to me.

**Comment by adonath** [comment]
@AtreyeeS You could add one more comment in the tutorial to mention that the in future we will support varying energy thresholds.

**Pull request merged**